src/metabase/api/permission_graph.clj

```diff
@@ -7,105 +7,114 @@
-                            :identity ::table-graph))
-
-;; {:groups {1 {:data {:schemas {"PUBLIC" ::schema-perms}}}}}
-(s/def ::schema-graph (s/map-of ::schema-name ::schema-perms
-                                :conform-keys true))
-
-;; {:groups {1 {:data {:schemas ::schemas}}}}
-(s/def ::schemas (s/or :str->kw   #{"all" "segmented" "none" "block" "full" "limited"}
-                       :nil->none nil?
-                       :identity  ::schema-graph))
-
-(s/def ::data (s/keys :opt-un [::native ::schemas]))
-
-(s/def ::download (s/keys :opt-un [::native ::schemas]))
-
-(s/def ::data-model (s/keys :opt-un [::native ::schemas]))
-
-;; We use "yes" and "no" instead of booleans for consistency with the application perms graph, and consistency with the
-;; language used on the frontend.
-(s/def ::details (s/or :str->kw #{"yes" "no"}))
-
-(s/def ::db-perms (s/keys :opt-un [::data ::download ::data-model ::details ::execute]))
-
-(s/def ::db-graph
-  (s/map-of ::id
-            ::db-perms
-            :conform-keys true))
-
-(s/def :metabase.api.permission-graph.data/groups
-  (s/map-of ::id ::db-graph
-            :conform-keys true))
-
-(s/def ::data-permissions-graph
-  (s/keys :req-un [:metabase.api.permission-graph.data/groups]))
+(def ^:private table-perms
+  [:or
+   [:enum :all :segmented :none :full :limited]
+   [:map
+    [:read [:enum :all :none]]
+    [:query [:enum :all :none :segmented]]]])
+
+(def ^:private schema-perms
+  [:or
+   [:keyword {:title "schema name"}]
+   [:map-of id table-perms]])
+
+(def ^:private schema-graph
+  [:map-of
+   [:string {:decode/perm-graph name}]
+   schema-perms])
+
+(def ^:private schemas
+  [:or
+   [:enum :all :segmented :none :block :full :limited]
+   schema-graph])
+
+(def ^:private data-perms
+  [:map
+   [:native {:optional true} native]
+   [:schemas {:optional true} schemas]])
+
+(def strict-data-perms
+  "data perms that care about how native and schemas keys related to one another"
+  [:and
+   data-perms
+   [:fn {:error/fn (constantly
+                    (trs "Invalid DB permissions: If you have write access for native queries, you must have full data access."))}
+    (fn [{:keys [native schemas]}]
+      (not (and (= native :write) schemas (not= schemas :all))))]])
+
+(def ^:private db-graph
+  [:map-of
+   id
+   [:map
+    [:data {:optional true} data-perms]
+    [:download {:optional true} data-perms]
+    [:data-model {:optional true} data-perms]
+    ;; "We use :yes and :no instead of booleans for consistency with the application perms graph, and"
+    ;; "consistency with the language used on the frontend."
+    [:details {:optional true} [:enum :yes :no]]
+    [:execute {:optional true} [:enum :all :none]]]])
+
+(def strict-db-graph
+  "like db-graph, but strict"
+  (-> db-graph
+      (mut/assoc-in [1 :data] strict-data-perms)
+      (mut/assoc-in [1 :download] strict-data-perms)
+      (mut/assoc-in [1 :data-model] strict-data-perms)
```

Comment:
I wonder if it would be clearer to just duplicate the `db-graph` code here using `strict-data-perms`? It might make it more obvious what's going on, and it's not that much more code anyway. Idk, just a thought.

Reply:
Good call, I got kind of wrapped up on how easy it is to modify schemas (since they're isomorphic with data)